cocotb/runner.py

```diff
@@ -0,0 +1,807 @@
+        self.build_dir = os.path.abspath(build_dir)
+        os.makedirs(self.build_dir, exist_ok=True)
+
+        # note: to avoid mutating argument defaults, we ensure that no value
+        # is written without a copy. This is much more concise and leads to
+        # a better docstring than using `None` as a default in the parameters
+        # list.
+        self.library_name = library_name
+        self.verilog_sources = get_abs_paths(verilog_sources)
+        self.vhdl_sources = get_abs_paths(vhdl_sources)
+        self.includes = get_abs_paths(includes)
+        self.defines = list(defines)
+        self.parameters = dict(parameters)
+        self.compile_args = list(extra_args)
+        self.always = always
+        self.hdl_toplevel = toplevel
+
+        for e in os.environ:
+            self.env[e] = os.environ[e]
+
+        cmds = self.build_command()
+        self.execute(cmds, cwd=self.build_dir)
+
+    def test(
+        self,
+        py_module: Union[str, Sequence[str]],
+        toplevel: str,
+        toplevel_lang: Optional[str] = None,
+        testcase: Optional[str] = None,
+        seed: Optional[Union[str, int]] = None,
+        python_search: Sequence[PathLike] = [],
+        extra_args: Sequence[str] = [],
+        plus_args: Sequence[str] = [],
+        extra_env: Mapping[str, str] = {},
+        waves: Optional[bool] = None,
+        gui: Optional[bool] = None,
+        sim_dir: Optional[PathLike] = None,
+    ) -> PathLike:
+        """Run a test."""
+
+        if sim_dir is None:
+            self.sim_dir = self.build_dir
+        else:
+            self.sim_dir = os.path.abspath(sim_dir)
+
+        if isinstance(py_module, str):
+            self.module = py_module
+        else:
+            self.module = ",".join(py_module)
+
+        # note: to avoid mutating argument defaults, we ensure that no value
+        # is written without a copy. This is much more concise and leads to
+        # a better docstring than using `None` as a default in the parameters
+        # list.
+        self.python_search = list(python_search)
+        self.sim_toplevel = toplevel
+        self.toplevel_lang = self.check_toplevel_lang(toplevel_lang)
+        self.sim_args = list(extra_args)
+        self.plus_args = list(plus_args)
+        self.env = dict(extra_env)
+
+        if testcase is not None:
+            self.env["TESTCASE"] = testcase
+
+        if seed is not None:
+            self.env["RANDOM_SEED"] = str(seed)
+
+        if waves is None:
+            self.waves = bool(int(os.getenv("COCOTB_WAVES", 0)))
+        else:
+            self.waves = bool(waves)
+
+        if gui is None:
+            self.gui = bool(int(os.getenv("COCOTB_GUI", 0)))
+        else:
+            self.gui = bool(gui)
+
+        results_xml_file = os.getenv(
+            "COCOTB_RESULTS_FILE", os.path.join(self.build_dir, "results.xml")
+        )
+
+        try:
+            os.remove(results_xml_file)
+        except OSError:
+            pass
```

Comment:
```suggestion
        with suppress(OSError):
            os.remove(results_xml_file)
```
(needs `from contextlib import suppress`)